operator/webhooks/pod_mutating_webhook.go

```diff
@@ -148,8 +158,104 @@ func (a *PodMutatingWebhook) isKeptnAnnotated(pod *corev1.Pod) (bool, error) {
+		if rsOwner.UID == "" {
+			return false, nil
+		}
+
+		dp := &appsv1.Deployment{}
+		return a.fetchParentObjectAndCopyLabels(ctx, rsOwner.Name, req.Namespace, pod, dp)
+	case "StatefulSet":
+		sts := &appsv1.StatefulSet{}
+		return a.fetchParentObjectAndCopyLabels(ctx, podOwner.Name, req.Namespace, pod, sts)
+	case "DaemonSet":
+		ds := &appsv1.DaemonSet{}
+		return a.fetchParentObjectAndCopyLabels(ctx, podOwner.Name, req.Namespace, pod, ds)
+	default:
+		return false, nil
+	}
+}
+
+func (a *PodMutatingWebhook) fetchParentObjectAndCopyLabels(ctx context.Context, name string, namespace string, pod *corev1.Pod, objectContainer client.Object) (bool, error) {
+	if err := a.Client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, objectContainer); err != nil {
+		return false, nil
+	}
+	objectContainerMetaData := metav1.ObjectMeta{
+		Labels:      objectContainer.GetLabels(),
+		Annotations: objectContainer.GetAnnotations(),
+	}
+	return a.copyResourceLabelsIfPresent(&objectContainerMetaData, pod)
+}
+
+func (a *PodMutatingWebhook) copyResourceLabelsIfPresent(sourceResource *metav1.ObjectMeta, targetPod *corev1.Pod) (bool, error) {
+	var workloadName, appName, version, preDeploymentChecks, postDeploymentChecks, preEvaluationChecks, postEvaluationChecks string
+	var gotWorkloadName, gotAppName, gotVersion, gotPreDeploymentChecks, gotPostDeploymentChecks, gotPreEvaluationChecks, gotPostEvaluationChecks bool
+
+	workloadName, gotWorkloadName = getLabelOrAnnotation(sourceResource, common.WorkloadAnnotation, common.K8sRecommendedWorkloadAnnotations)
+	appName, gotAppName = getLabelOrAnnotation(sourceResource, common.AppAnnotation, common.K8sRecommendedAppAnnotations)
+	version, gotVersion = getLabelOrAnnotation(sourceResource, common.VersionAnnotation, common.K8sRecommendedVersionAnnotations)
+	preDeploymentChecks, gotPreDeploymentChecks = getLabelOrAnnotation(sourceResource, common.PreDeploymentTaskAnnotation, "")
+	postDeploymentChecks, gotPostDeploymentChecks = getLabelOrAnnotation(sourceResource, common.PostDeploymentTaskAnnotation, "")
+	preEvaluationChecks, gotPreEvaluationChecks = getLabelOrAnnotation(sourceResource, common.PreDeploymentEvaluationAnnotation, "")
+	postEvaluationChecks, gotPostEvaluationChecks = getLabelOrAnnotation(sourceResource, common.PostDeploymentEvaluationAnnotation, "")
+
+	if len(workloadName) > common.MaxWorkloadNameLength || len(version) > common.MaxVersionLength {
+		return false, ErrTooLongAnnotations
+	}
+
+	if len(targetPod.Annotations) == 0 {
+		targetPod.Annotations = make(map[string]string)
+	}
+
+	if gotWorkloadName {
+		targetPod.Annotations[common.WorkloadAnnotation] = workloadName
+
+		if !gotVersion {
+			targetPod.Annotations[common.VersionAnnotation] = a.calculateVersion(targetPod)
+		} else {
+			targetPod.Annotations[common.VersionAnnotation] = version
+		}
+
+		if gotAppName {
+			targetPod.Annotations[common.AppAnnotation] = appName
+		}
+
+		if gotPreDeploymentChecks {
```

Comment:
maybe we can reduce a bit of complexity here with a loop on a list of annotations?